src/Workspaces/Remote/Core/SolutionAssetStorage.cs

```diff
@@ -13,121 +13,115 @@
+            : await solutionState.GetChecksumAsync(projectId, cancellationToken).ConfigureAwait(false);
 
-                var scope = new Scope(this, checksum, solutionInfo);
-                refCountAndScope = (refCount: 1, scope);
-                _checksumToRefCountedScope[checksum] = refCountAndScope;
+        lock (_gate)
+        {
+            if (_checksumToScope.TryGetValue(checksum, out var scope))
+            {
+                Contract.ThrowIfTrue(scope.RefCount <= 0);
+                scope.RefCount++;
                 return scope;
             }
-        }
 
-        private void DisposeScope(Scope scope)
-        {
-            SolutionReplicationContext replicationContext;
+            var solutionInfo = new PinnedSolutionInfo(
+                checksum,
+                fromPrimaryBranch: solutionState.BranchId == solutionState.Workspace.PrimaryBranchId,
+                solutionState.WorkspaceVersion);
 
-            lock (_gate)
-            {
-                var checksum = scope.Checksum;
-                var (refCount, existingScope) = _checksumToRefCountedScope[checksum];
-                Contract.ThrowIfTrue(existingScope != scope);
+            scope = new Scope(this, checksum, solutionInfo, solutionState);
+            _checksumToScope[checksum] = scope;
+            return scope;
+        }
+    }
 
-                Contract.ThrowIfTrue(refCount <= 0);
-                refCount--;
+    private void DecreaseScopeRefCount(Scope scope)
+    {
+        lock (_gate)
+        {
+            var checksum = scope.Checksum;
+            var existingScope = _checksumToScope[checksum];
+            Contract.ThrowIfTrue(existingScope != scope);
 
-                // If our refcount is still above 0, then just update the map and return.  Nothing else to do at this point.
-                if (refCount > 0)
-                {
-                    _checksumToRefCountedScope[checksum] = (refCount, scope);
-                    return;
-                }
+            Contract.ThrowIfTrue(scope.RefCount <= 0);
+            scope.RefCount--;
 
-                // Last ref went away, update our maps while under the lock, then cleanup its context data outside of the lock.
-                _checksumToRefCountedScope.Remove(checksum);
-                Contract.ThrowIfFalse(_solutionStates.TryRemove(scope.SolutionInfo.SolutionChecksum, out var stateAndContext));
-                replicationContext = stateAndContext.ReplicationContext;
-            }
+            // If our refcount is still above 0, then nothing else to do at this point.
+            if (scope.RefCount > 0)
+                return;
 
-            replicationContext.Dispose();
+            // Last ref went away, update our maps while under the lock, then cleanup its context data outside of the lock.
+            _checksumToScope.Remove(checksum);
         }
 
+        scope.ReplicationContext.Dispose();
+    }
+
+    internal partial class Scope
```

Comment:
these helpers moved to be isntance methods on Scope.  They weren't using anything else.   SolutionAssetStorage is basically just a holder for the inflight operations (scopes) for a particular solution checksum.

Reply:
i did this inline in this file with a partial tyepe to make the diff easy to read.  if you are ok with this @dibarbet  i will move this code into the other part.

Reply:
yup makes sense to me!